admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,191 @@
+        const trajectory_msgs::msg::JointTrajectoryPoint &reference_joint_state,
+        const rclcpp::Duration &period,
+        trajectory_msgs::msg::JointTrajectoryPoint &desired_joint_states);
+
+    void get_controller_state(control_msgs::msg::AdmittanceControllerState &state_message);
+
+
+  public:
+    // admittance config parameters
+    std::shared_ptr<admittance_struct_parameters::admittance_struct> parameter_handler_;
+    admittance_struct_parameters::admittance_struct::params *parameters_;
+    admittance_struct_parameters::admittance_struct::params parameters_copy;
+
+  protected:
+    /**
+     * All values are in the controller frame
+     */
+    void calculate_admittance_rule(
+        const Eigen::Matrix<double, 3, 2> &wrench,
+        const Eigen::Matrix<double, 3, 2> &desired_vel,
+        const double dt
+    );
+
+    void process_wrench_measurements(
+        const geometry_msgs::msg::Wrench &measured_wrench, const Eigen::Matrix<double, 3, 3> &sensor_rot,
+        const Eigen::Matrix<double, 3, 3> &cog_rot
+    );
+
+    Eigen::Vector3d get_rotation_axis(const Eigen::Matrix3d &R) const;
+
+    void convert_cartesian_deltas_to_joint_deltas(const std::vector<double> &positions,
+                                                  const Eigen::Matrix<double, 3, 2> &cartesian_delta,
+                                                  const std::string &link_name,
+                                                  Eigen::Matrix<double, Eigen::Dynamic, 1> &joint_delta, bool &success);
+
+    Eigen::Matrix<double, 3, 2> convert_joint_deltas_to_cartesian_deltas(const std::vector<double> &positions,
+                                                                         const std::vector<double> &joint_delta,
+                                                                         const std::string &link_name,
+                                                                         bool &success);
+
+    void normalize_rotation(Eigen::Matrix<double, 3, 3, Eigen::ColMajor> &R);
+
+    Eigen::Matrix<double, 4, 4, Eigen::ColMajor>
+    get_transform(const std::vector<double> &positions, const std::string &link_name, bool external, bool &success);
+
+    void eigen_to_msg(const Eigen::Matrix<double, 3, 2> &wrench, const std::string &frame_id,
+                      geometry_msgs::msg::WrenchStamped &wrench_msg);
+
+    template<typename T1, typename T2>
+    void vec_to_eigen(const std::vector<T1> data, T2 &matrix);
+
+    // Kinematics interface plugin loader
+    std::shared_ptr<pluginlib::ClassLoader<kinematics_interface::KinematicsBaseClass>> kinematics_loader_;
+    std::unique_ptr<kinematics_interface::KinematicsBaseClass> kinematics_;
+
+    // number of robot joint
+    int num_joints_;
+
+    // buffers to pass data to kinematics interface
+    std::vector<double> transform_buffer_vec_;
+    std::vector<double> joint_buffer_vec_;
+    std::vector<double> cart_buffer_vec_;
+    std::vector<double> jacobian_buffer_vec_;
```

Comment:
1. `transform_buffer_vec_` is 16 doubles and is used to pass tranformation from the kinematics interface to the controller. Technically, this could be stack allocated, but `std::vector` can't put its memory on the  stack as far as I known.
2. `joint_buffer_vec_` is an unknown length (num of joints) vector of doubles used to pass values from the  controller to the kinematics interface. This needs to be a member because dynamic allocation is needed.
3. `cart_buffer_vec_` is used to pass admittance velocities and accelerations to the kinematics IK. This is a fixed size of 6, but we again have the `std::vector` issue.    
4. `jacobian_buffer_vec_` is unused and needs to be deleted.

Reply:
We have to adjust interface to `kinematics_interface` having a lot of buffers doesn't seem like a good design.